accounts/manager.go

```diff
@@ -34,8 +35,34 @@ type Manager struct {
 
 	feed event.Feed // Wallet feed notifying of arrivals/departures
 
-	quit chan chan error
-	lock sync.RWMutex
+	quit      chan chan error
+	lock      sync.RWMutex
+	nonceLock accountNonceLocker // Used to lock account nonces, to prohibit nonce reuse
+}
+
+type accountNonceLocker struct {
+	mut   sync.Mutex
+	locks map[common.Address]*sync.Mutex
+}
+
+// LockNonce Locks a particular nonce for an account, to prevent another tx getting the same nonce until this lock has been released
+// This should be used when a nonce has not been provided by the user,
+// and we get a nonce from the pools. The mutex prevents the (an identical nonce) from being
+// read again during the time that the first transaction is being signed.
+func (am *Manager) LockNonce(address common.Address) {
+	if _, exist := am.nonceLock.locks[address]; !exist {
+		am.nonceLock.mut.Lock()
+		if _, exist := am.nonceLock.locks[address]; !exist {
```

Comment:
So, it's done twice so that in 99% of all cases, there's no need to even obtain the lock (if the noncelock for that address already exists, it will exit at the first check). 

If the noncelock does not exist, it needs to lock the `mut`. And at that point, it needs to check again, because there may have been  two threads doing that unsynchronized first check and now about to create the address-lock. 

I don't want to use defer, because there's a tricky thing here, regarding scoping. I'm not sure when the deferred method is executed, but if it's at the end, I don't think it's very nice. 

This is how I imagine a deferred will look like without syntactic sugar: 

```golang
func (am *Manager) LockNonce(address common.Address) {
	if _, exist := am.nonceLock.locks[address]; !exist {
		am.nonceLock.mut.Lock()
		if _, exist := am.nonceLock.locks[address]; !exist {
			am.nonceLock.locks[address] = new(sync.Mutex)
		}
		
	}
	am.nonceLock.locks[address].Lock()
	am.nonceLock.mut.Unlock()
	return
}
``` 

The problem with that is that it will try to obtain the _addresslock_ before having released the `nonceLock`. So if some other process already has the _addresslock_, this will stall other callers trying to lock the nonce for totally separate addresses, since they will be unable to obtain `mut`. 

Course, I may be wrong about how `defer` works (?)